In [1]:
import pandas as pd
import numpy as np

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [25]:
#task 1
df = pd.read_csv("/content/drive/MyDrive/DAV_LAB01/adult.csv")

print("INSPECTING FULL SUMMARY \n")
df.info()
6
print("\n Head 5 rows")
print(df.head())

print("\n Missing entries check")
print(df.isnull().sum())

print("\n After replacement with '?'\n ")
df.replace('?', np.nan , inplace = True)
print(df.isnull().sum())


INSPECTING FULL SUMMARY 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB

 Head 5 rows
   age workclass  fnlwgt     education  education.num

In [31]:
#task 2
print("MISSING VALUES : \n")
print(df.isnull().sum().sort_values(ascending = False))

print("\n PERCENTAGE OF MISSING VALUES : \n")
print((df.isnull().mean()*100).sort_values(ascending = False))

print("\n Occupation has most missing data")

MISSING VALUES : 

occupation        1843
workclass         1836
native.country     583
fnlwgt               0
education            0
education.num        0
age                  0
marital.status       0
relationship         0
sex                  0
race                 0
capital.gain         0
capital.loss         0
hours.per.week       0
income               0
dtype: int64

 PERCENTAGE OF MISSING VALUES : 

occupation        5.660146
workclass         5.638647
native.country    1.790486
fnlwgt            0.000000
education         0.000000
education.num     0.000000
age               0.000000
marital.status    0.000000
relationship      0.000000
sex               0.000000
race              0.000000
capital.gain      0.000000
capital.loss      0.000000
hours.per.week    0.000000
income            0.000000
dtype: float64

 Occupation has most missing data


In [71]:
# Task 3

print("""
workclass replace missing values with 'Unknown' because around 5.64% of
the values are missing. This preserves the observations and avoids
incorrectly assigning the most common workclass.

occupation replace missing values with 'Unknown' because around 5.66% of
the values are missing this preserves the observations and avoids
incorrectly assigning the most common occupation.
""")

df['workclass'] = df['workclass'].fillna('Unknown')
df['occupation'] = df['occupation'].fillna('Unknown')
print(df.isnull().sum())


workclass replace missing values with 'Unknown' because around 5.64% of
the values are missing. This preserves the observations and avoids
incorrectly assigning the most common workclass.

occupation replace missing values with 'Unknown' because around 5.66% of
the values are missing this preserves the observations and avoids
incorrectly assigning the most common occupation.

age                 0
workclass           0
fnlwgt              0
education           0
education.num       0
marital.status      0
occupation          0
relationship        0
race                0
sex                 0
capital.gain        0
capital.loss        0
hours.per.week      0
native.country    556
income              0
dtype: int64


In [72]:
#task 4
print("dup rows: ")
print(df.duplicated().sum())

features = df.drop(columns=['income'])
feature_duplicates = features.duplicated().sum()
print("Duplicates ignoring income:", feature_duplicates)

print("\nRows duplicated based on features:")
print(df[features.duplicated(keep=False)])

print("I would not consider them duplicate records just because their features are identical. Two different people can have the same age, education, occupation and other characteristics but different income outcomes.")

dup rows: 
24
Duplicates ignoring income: 25

Rows duplicated based on features:
       age         workclass  fnlwgt     education  education.num  \
4963    39           Private  138192     Bachelors             13   
6227    90           Private   52386  Some-college             10   
7615    19           Private  251579  Some-college             10   
7978    25           Private  308144     Bachelors             13   
8356    21           Private  250051  Some-college             10   
8453    25           Private  308144     Bachelors             13   
8500    38           Private  207202       HS-grad              9   
8645    90           Private   52386  Some-college             10   
9067    27           Private  255582       HS-grad              9   
9269    20           Private  107658  Some-college             10   
9305    25           Private  195994       1st-4th              2   
10307   19           Private  138153  Some-college             10   
10357   49  Self-emp-n

In [42]:
#task 5
print("Education Unique values : " , df['education'].unique())
print("\n Marital status unique values : " , df['marital.status'].unique())
print("\n Native country unique values : " , df['native.country'].unique())

print("NO TRAILING SPACES")

Education Unique values :  ['HS-grad' '7th-8th' 'Some-college' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

 Marital status unique values :  ['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

 Native country unique values :  ['United-States' nan 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'Hungary' 'Columbia' 'Guatemala'
 'El-Salvador' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Portugal' 'Laos'
 'Thailand' 'Outlying-US(Guam-USVI-etc)' 'Scotland']


In [48]:
# check for trailing spaces
print("Education:")
for value in df['education'].unique():
    print(repr(value))

print("\nMarital-status:")
for value in df['marital.status'].unique():
    print(repr(value))

print("\nNative-country:")
for value in df['native.country'].unique():
    print(repr(value))

print("Education whitespace:",
      df['education'].astype(str).str.len().ne(
          df['education'].astype(str).str.strip().str.len()
      ).sum())

print("Marital-status whitespace:",
      df['marital.status'].astype(str).str.len().ne(
          df['marital.status'].astype(str).str.strip().str.len()
      ).sum())

print("Native-country whitespace:",
      df['native.country'].astype(str).str.len().ne(
          df['native.country'].astype(str).str.strip().str.len()
      ).sum())


#df['education'] = df['education'].str.strip()

Education:
'HS-grad'
'7th-8th'
'Some-college'
'10th'
'Doctorate'
'Prof-school'
'Bachelors'
'Masters'
'11th'
'Assoc-acdm'
'Assoc-voc'
'1st-4th'
'5th-6th'
'12th'
'9th'
'Preschool'

Marital-status:
'Widowed'
'Divorced'
'Separated'
'Never-married'
'Married-civ-spouse'
'Married-spouse-absent'
'Married-AF-spouse'

Native-country:
'United-States'
nan
'Mexico'
'Greece'
'Vietnam'
'China'
'Taiwan'
'India'
'Philippines'
'Trinadad&Tobago'
'Canada'
'South'
'Holand-Netherlands'
'Puerto-Rico'
'Poland'
'Iran'
'England'
'Germany'
'Italy'
'Japan'
'Hong'
'Honduras'
'Cuba'
'Ireland'
'Cambodia'
'Peru'
'Nicaragua'
'Dominican-Republic'
'Haiti'
'Hungary'
'Columbia'
'Guatemala'
'El-Salvador'
'Jamaica'
'Ecuador'
'France'
'Yugoslavia'
'Portugal'
'Laos'
'Thailand'
'Outlying-US(Guam-USVI-etc)'
'Scotland'
Education whitespace: 0
Marital-status whitespace: 0
Native-country whitespace: 0


In [56]:
#task 6
print("NUMERIC COLS " , df.describe())

print("\nCATEGORICAL COLS " ,df.describe(include='object'))

print("\n MEAN AGE : " , df['age'].mean())
print("\n MEDIAN AGE : " , df['age'].median())

print("\n Mean hours per week", df['hours.per.week'].mean())
print("\n Median hours per week", df['hours.per.week'].median())

print("The mean and median for both age and hours/week are relatively close age has a mean of 38.44 and median of 37 while hours/week has a mean of 40.95 and median of 40 the small differences suggest that neither variable is heavily affected by extreme values but age shows slightly more right skewness")

NUMERIC COLS                  age        fnlwgt  education.num  capital.gain  capital.loss  \
count  30725.000000  3.072500e+04   30725.000000  30725.000000  30725.000000   
mean      38.439512  1.898538e+05      10.129699   1105.785094     88.889959   
std       13.119665  1.054585e+05       2.562702   7497.027762    405.613208   
min       17.000000  1.376900e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178330e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.785170e+05      10.000000      0.000000      0.000000   
75%       47.000000  2.373170e+05      13.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    30725.000000  
mean        40.946461  
std         11.987385  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000  

CATEGORICAL COLS         workclass education    

In [62]:
#task 7
print("most common occupation")
print(df['occupation'].value_counts())
#print(df['occupation'].value_counts().idxmax())

print("\npercentage distribution of sex")
print(df['sex'].value_counts(normalize=True)*100)


print("\n percentage distribution of the income lable")
print(df['income'].value_counts(normalize=True)*100)
#The income classes are imbalanced

most common occupation
occupation
Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         3770
Sales                3650
Other-service        3295
Machine-op-inspct    2002
Transport-moving     1597
Handlers-cleaners    1370
Farming-fishing       994
Tech-support          928
Protective-serv       649
Priv-house-serv       149
Armed-Forces            9
Unknown                 7
Name: count, dtype: int64

percentage distribution of sex
sex
Male      67.674532
Female    32.325468
Name: proportion, dtype: float64

 percentage distribution of the income lable
income
<=50K    75.101709
>50K     24.898291
Name: proportion, dtype: float64


In [67]:
# Task 8
print("education and education-num consistency")
print(df.groupby('education')['education.num'].unique())
education_mapping = df.groupby('education')['education.num'].unique()

print("No inconsistencies were found")

education and education-num consistency
education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object
No inconsistencies were found


In [70]:
#task 9
import sqlite3

df_original = pd.read_csv("/content/drive/MyDrive/DAV_LAB01/adult.csv")
print("Original CSV shape:", df_original.shape)


#local SQLite database
connection = sqlite3.connect("adult_income.db")

df_original.to_sql(
    "adult_income",
    connection,
    if_exists="replace",
    index=False
)
print("CSV data successfully loaded into SQLite table.")

query = """
SELECT *
FROM adult_income
WHERE age > 30;
"""

df_sql = pd.read_sql_query(query, connection)

print("\nSQL query result:")
print(df_sql.head())
print("\nSQL result shape:", df_sql.shape)

print("\nOriginal CSV shape:", df_original.shape)
print("SQL extracted shape:", df_sql.shape)

connection.close()

Original CSV shape: (32561, 15)
CSV data successfully loaded into SQLite table.

SQL query result:
   age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0  

## Data Profiling Summary Report

### Dataset overview

The dataset contains 32561 rows and 15 columns. It is the Adult Income dataset and contains information about people such as their age, workclass, education, occupation, marital status, relationship, sex, hours worked per week and income. The income column shows whether a person earns <=50K or >50K.

### Data quality issues found

The main data quality issue was missing values represented by the character ?. After replacing ? with NaN, missing values were found in workclass, occupation and native country. Workclass had 1836 missing values, occupation had 1843 and native country had 583. Duplicate rows were also checked using duplicated. Education and education num were checked to make sure that each education level always had the same numerical value and no inconsistencies were found.

### Key observations

The most common occupation was Prof specialty. The mean age was about 38 years while the median was 37 years, so there was only a small difference between them. The mean hours per week was about 41 while the median was 40. The income classes were imbalanced because around 76 percent of the records belonged to the <=50K class while around 24 percent belonged to the >50K class. The dataset therefore contains many more lower income records than higher income records.

### Cleaning decisions made

The ? values were first replaced with proper NaN values so that they could be correctly identified as missing data. Missing values in workclass, occupation and native country were replaced with Unknown. This was done to keep the records instead of removing them and to avoid incorrectly guessing the missing categories. Leading and trailing spaces were removed from categorical columns using strip. Education and education num were also checked and found to be consistent with each other.
